brille/design/Proposed_integrated_syntax_Brill_SpinW_Horace.md

```diff
@@ -0,0 +1,267 @@
+```
+
+And the proposed Python syntax:
+
+<table>
+<td><img src="https://placehold.it/5x150/000000/000000?text=+"></td>
+<td>
+<pre lang="python">
+import brille as b
+tri = h.sw_model('triAF', 1)
+spinw_pars = {'hermit': False, 'formfact': True, 'resfun': 'gauss'}
+brl_pars = {'max_volume':1e-5}
+brl = b.Brille(tri, **brl_pars)
+sim1 = h.disp2sqw_eval(cut1, brl.horace_sqw, fwhm, **spinw_pars)
+</pre>
+</td>
+<td><img src="https://placehold.it/5x150/000000/000000?text=+"></td>
+</table>
+
+This syntax assumes that the Horace python distribution (`pyHorace`) will include SpinW as well as Horace.
+This would enable both to use the same Matlab interpreter process, and make intercommunication much easier and faster.
+Again these syntax should work for powder data objects, possibly replacing `h.disp2sqw` with `m.disp2sqw` method of `MSlice` (yet to be implemented).
+
+
+## Fitting the data with resolution convolution
+
+Horace includes a fitting framework which uses a Levenberg-Marquardt minimizer.
+The following example uses Brille and the TobyFit Monte Carlo resolution convolution method:
+
+```matlab
+proj = projaxes([1 0 0], [0 1 0]);
+cut1 = cut_sqw(sqw_file, proj, [hmin hstep hmax], [kmin kstep kmax], [lmin lmax], [emin estep emax]);
+model = pcsmo(JF1, JA, JF2, JF3, Jperp, D);   % User model defined as a function
+model_pars = {'mat', {'JF1', 'JA', 'JF2', 'JF3', 'Jperp', 'D(3,3)'}, ...
+              'hermit', false, 'optmem', 0, 'useFast', true, 'formfact', true, ...
+              'resfun', 'gauss', 'coordtrans', diag([2 2 1 1])};
+brl_pars = {'max_volume', 1e-5};
+brl = brillem.Brille(model, brl_pars{:});
+sample = IX_sample(true,[0,0,1],[0,1,0],'cuboid',[0.01,0.05,0.01]);
+maps = maps_instrument(ei, freq, 'S');
+cut1 = set_sample(cut1, sample);
+cut1 = set_instrument(cut1, maps);
+tbf = tobyfit(cut1);
+tbf = tbf.set_fun (@brl.horace_sqw, {[J1 J2 J3 J4 J5 J6 0.1] model_pars{:}});
+tbf = tbf.set_mc_points(5);
+[data_out, fitdata] = tbf.fit();
+```
+
+The proposed Python syntax for this is:
+
+<table>
+<td><img src="https://placehold.it/5x400/000000/000000?text=+"></td>
+<td>
+<pre lang="python">
+from pyHorace import horace as h
+import brille as b
+<br>
+proj = {'u':[1,0,0], 'v':[0,1,0], 'type':'rrr'}
+cut1 = h.cut_sqw(sqw_file, proj, [hmin, hstep, hmax], [kmin, kstep, kmax], 
+                                 [lmin, lmax], [emin, estep, emax])
+model = pcsmo(JF1, JA, JF2, JF3, Jperp, D)   # User model defined as a function
+model_pars = {'mat':['JF1', 'JA', 'JF2', 'JF3', 'Jperp', 'D(3,3)'],
+              'hermit':False, 'optmem', 0, 'useFast':True, 'formfact':True,
+              'resfun':'gauss', 'coordtrans':np.diag([2 2 1 1])}
+brl_pars = {'max_volume', 1e-5};
+brl = b.Brille(model, brl_pars{:});
+sample = h.IX_sample(True,[0,0,1],[0,1,0],'cuboid',[0.01,0.05,0.01]);
+maps = h.maps_instrument(ei, freq, 'S');
+cut1.set_sample(sample);
+cut1.set_instrument(maps);
+tbf = tobyfit(cut1);
```

Comment:
As a general rule I'm not a fan of the design pattern:
```python
object = MyClass()
object.set(**essential_data)
object.do_the_thing()
```
where some elegant alternatives are `MyClass(**essential_data).do_the_thing()` or even `do_the_thing(**essential_data)`.
The problem is that these workflows often allow one to construct objects in a "useless" state. This state needs to be tested, methods should fail elegantly if the required `set()` has not been performed yet, and users need to be taught to follow a sequence of actions rather than satisfying the input arguments of a single action.

I'm not very familiar with this tool/workflow, so it may be that there is a good reason that tobyfit is instantiated before setting properties and then fitting is called separately. But as I see it is imitating the Matlab workflow, it seems worth reviewing! Perhaps at least some steps would make sense to bundle together? Which steps are useful to separate (e.g. to perform multiple fitting variations after a slow data-loading step?)

Reply:
Yeah, this is a bit tricky... We (e.g. Toby) came to the decision that the `sample` and `instrument` should be properties of the data object rather than the fit object since it can be passed around multiple fit objects; however it's not essential to look at the data (only essential for modelling it) so it should not be included in the constructor for the `sqw` object `cut1` (e.g. in the `cut_sqw` function). 

However, it *is* essential for the resolution convolution modelling, so maybe it should be included in the `tobyfit` constructor instead of in the data object... In which case the above calls would look like:

```python
tbf = h.tobyfit(cut1,
                h.IX_sample(True,[0,0,1],[0,1,0],'cuboid',[0.01,0.05,0.01]);
                h.maps_instrument(ei, freq, 'S'))
```